File: tfjs-backend-wasm/src/cc/fft_impl.cc
/* Copyright 2020 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ===========================================================================*/

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#endif

#include <cmath>
#include <vector>

#include "src/cc/backend.h"
#include "src/cc/fft_impl.h"

namespace tfjs {
namespace wasm {

void fft(const size_t real_input_id, const size_t imag_input_id,
         const size_t outer_dim, const size_t inner_dim,
         const size_t is_real_component, const size_t out_id) {
  auto& real_input_info = backend::get_tensor_info(real_input_id);
  const float* real_input_buf = real_input_info.f32();
  auto& imag_input_info = backend::get_tensor_info(imag_input_id);
  const float* imag_input_buf = imag_input_info.f32();

  auto& out_info = backend::get_tensor_info_out(out_id);
  float* out_buf_ptr = out_info.f32_write();
  const size_t input_size = real_input_info.size;

  const float exponent_multiplier = -2.0 * M_PI;

  for (size_t row = 0; row < outer_dim; ++row) {
    for (size_t col = 0; col < inner_dim; ++col) {
      float index_ratio = float(col) / float(inner_dim);
      float exponent_multiplier_times_index_ratio =
          exponent_multiplier * index_ratio;

      float result = 0.0;

      for (size_t i = 0; i < inner_dim; ++i) {
        float x = exponent_multiplier_times_index_ratio * float(i);
        float exp_r = cos(x);
        float exp_i = sin(x);
        float real = real_input_buf[row * inner_dim + i];
        float imag = imag_input_buf[row * inner_dim + i];

        if (is_real_component > 0) {
          result += real * exp_r - imag * exp_i;
        } else {
          result += real * exp_i + imag * exp_r;
        }
      }

      *out_buf_ptr = result;
      out_buf_ptr++;
    }
  }
}
}  // namespace wasm
}  // namespace tfjs

File: tfjs-backend-webgl/src/kernels/FFT_impl.ts
/**
 * @license
 * Copyright 2020 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */

import {TensorInfo, util} from '@tensorflow/tfjs-core';

import {MathBackendWebGL} from '../backend_webgl';
import {FFTProgram} from '../fft_gpu';

import {complex} from './Complex';
import {reshape} from './Reshape';

export function fftImpl(
    x: TensorInfo, inverse: boolean, backend: MathBackendWebGL): TensorInfo {
  const xData = backend.texData.get(x.dataId);

  const inputSize = util.sizeFromShape(x.shape);
  // Collapse all outer dimensions to a single batch dimension.
  const innerDimensionSize = x.shape[x.shape.length - 1];
  const batch = inputSize / innerDimensionSize;

  const input2D = reshape(
      {inputs: {x}, backend, attrs: {shape: [batch, innerDimensionSize]}});

  const xShape = input2D.shape as [number, number];
  const realProgram = new FFTProgram('real', xShape, inverse);
  const imagProgram = new FFTProgram('imag', xShape, inverse);

  const inputs = [
    {
      dataId: xData.complexTensorInfos.real.dataId,
      dtype: xData.complexTensorInfos.real.dtype,
      shape: xShape
    },
    {
      dataId: xData.complexTensorInfos.imag.dataId,
      dtype: xData.complexTensorInfos.imag.dtype,
      shape: xShape
    }
  ];

  const realPart = backend.runWebGLProgram(realProgram, inputs, 'float32');
  const imagPart = backend.runWebGLProgram(imagProgram, inputs, 'float32');

  const complexOutput =
      complex({inputs: {real: realPart, imag: imagPart}, backend});

  backend.disposeIntermediateTensorInfo(realPart);
  backend.disposeIntermediateTensorInfo(imagPart);

  const complexOutputReshaped =
      reshape({inputs: {x: complexOutput}, backend, attrs: {shape: x.shape}});

  backend.disposeIntermediateTensorInfo(complexOutputReshaped);

  return complexOutputReshaped;
}
